src/protocol/ascii.rs

```diff
@@ -393,22 +470,55 @@ impl AsciiProtocol<Stream> {
             .map(|_| ())
     }
 
-    pub(super) fn delete(&mut self, key: &str) -> Result<bool, MemcacheError> {
-        check_key_len(key)?;
-        write!(self.reader.get_mut(), "delete {}\r\n", key)?;
-        self.reader.get_mut().flush()?;
-        self.reader
-            .read_line(|response| match MemcacheError::try_from(response) {
-                Ok(s) => {
-                    if s == "DELETED\r\n" {
-                        Ok(true)
-                    } else {
-                        Err(ServerError::BadResponse(Cow::Owned(s.into())).into())
+    pub(super) fn deletes<K: AsRef<str>, I: IntoIterator<Item = K>>(
+        &mut self,
+        keys: I,
+    ) -> Result<Vec<bool>, MemcacheError> {
+        let mut sent_count = 0;
+        {
+            let reader = self.reader.get_mut();
+            for k in keys.into_iter() {
+                let key = k.as_ref();
+                check_key_len(key)?;
+                write!(reader, "delete {}\r\n", key)?;
+                sent_count += 1;
+            }
+            // Flush now that all the requests have been written.
+            reader.flush()?;
+        }
+
+        // Receive all the responses. If there were errors, return the first.
+
+        let mut deleted_list = Vec::with_capacity(sent_count);
+        let mut error_list: Vec<MemcacheError> = Vec::new();
+        for _ in 0..sent_count {
+            let one_result = self
+                .reader
+                .read_line(|response| match MemcacheError::try_from(response) {
+                    Ok(s) => {
+                        if s == "DELETED\r\n" {
+                            Ok(true)
+                        } else {
+                            Err(ServerError::BadResponse(Cow::Owned(s.into())).into())
+                        }
                     }
-                }
-                Err(MemcacheError::CommandError(CommandError::KeyNotFound)) => Ok(false),
-                Err(e) => Err(e),
-            })
+                    Err(MemcacheError::CommandError(CommandError::KeyNotFound)) => Ok(false),
+                    Err(e) => Err(e),
+                });
+            match one_result {
+                Ok(deleted) => deleted_list.push(deleted),
+                Err(e) => error_list.push(e),
```

Comment:
If we see unrecoverable errors like `IOError`, `ParseErrors`, `ClientError`, `ServerError`, we should return here instead.

Reply:
Agreed.